### PR TITLE
Add explicit transactions

### PR DESCRIPTION
**Pull request by arv**
Starting to take shape. Will need more test fixes and then I'm off to start with the ObjC/Java side.

Towards https://github.com/rocicorp/replicache/issues/27

### CODE REVIEW

**Comment by arv** [comment]
Things that isn't clear.
 - Should the cli expose begin transaction?
 - Who owns the transactions map? Maybe the connection does?
 - mutator name and args are not plugged in yet.

File: db/db.go [hunk 1]
```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+func (tx *readTransaction) Close() error {
+	if tx.Closed() {
+		return ClosedError
+	}
+	tx.id = -1
+	removeTransaction(tx)
+	return nil
+}
+
+type writeTransaction struct {
```
Comment (aboodman) [inline db/db.go:386]: Why draw the distinction in the Go code between read and write transactions? I think they can all be write transactions and it can just be an API/typesystem thing whether it's a read or write tx.

I refreshed my memory and the cost of a `MapEditor` with no edits is basically zero. So there's no reason to avoid eagerly creating one.

It seems like you could eliminate half the code by literally just deleting the `readTransaction` type and wiring the callers to `writeTransaction`.

File: db/db.go [hunk 2]
```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &writeTransaction{txID, db, db.noms, db.head, db.head.Data(db.noms).Edit(), false}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) WriteTransaction(body func(tx WriteTransaction)) error {
```
Comment (aboodman) [inline db/db.go:298]: I think I would pull this and the corresponding `ReadTransaction` out of the `db` package just to make the API a little cleaner and easier to grok. Currently the only callers are test code. In that case just put a utility in a test file.

File: db/db.go [hunk 3]
```diff
@@ -26,6 +26,14 @@ const (
 	REMOTE_DATASET = "remote"
 )
 
+var (
+	transactions       = map[int]ReadTransaction{}
```
Comment (aboodman) [inline db/db.go:30]: Make these owned by `DB`.

File: db/db.go [hunk 4]
```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+
+	newMap := tx.me.Build()
+	newDataChecksum := newMap.NomsChecksum()
+	newData := tx.noms.WriteValue(newMap.NomsMap())
+
+	commit := makeTx(tx.noms, basis, time.DateTime(), function, args, newData, newDataChecksum)
+
+	ref = tx.noms.WriteValue(commit.Original)
+
+	// FastForward not strictly needed here because we should have already ensured that we were
```
Comment (aboodman) [inline db/db.go:518]: This comment is no longer correct. We are now relying on the internal optimistic locking of Noms again.

File: db/db.go [hunk 5]
```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+	// fast-forwarding outside of Noms, but it's a nice sanity check.
```
Comment (aboodman) [inline db/db.go:519]: It seems like there's a lot of code in e.g.., `execInternal` that needs to get removed as part of this PR.

File: db/db.go [hunk 6]
```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+	_, err = tx.noms.FastForward(tx.noms.GetDataset(LOCAL_DATASET), ref)
+	if err != nil {
+		return
+	}
+	tx.db.head = commit
```
Comment (aboodman) [inline db/db.go:524]: I'm not sure about this part, we need to think carefully here. This should be the only point we mutate the state in `DB`, and we need to think about what that means.

We need to update IIF the Noms commit succeeds. We don't want two write commits to interleave weirdly here and leave `db.head` pointing to an old head.

I think that really the introduction of transactions cleans up the concurrency story here a lot more. I think probably the right thing is to remove all the calls to `lock()` elsewhere and just have a single call here that wraps `FastForward` and the update to `tx.db.head` so that those are always atomic within this process.

If we ever want multiprocess support then we need to remove the `lock()` from the prev paragraph and instead Reload the database before each read (since it might have gotten changed by another process).

File: db/db.go [hunk 7]
```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+
 func (db *DB) lock() func() {
 	db.mu.Lock()
 	return func() {
 		db.mu.Unlock()
 	}
 }
+
+// ReadTransaction describes the operations allowed in a read transaction.
+type ReadTransaction interface {
```
Comment (aboodman) [inline db/db.go:313]: It's non-idiomatic to use interfaces this way in Go. Typically interfaces are at the caller of a boundary, not the callee.

Instead, prefer to expose the `readTransaction` struct (which might just become `transaction` -  see other comments).

File: repm/types.go [hunk 8]
```diff
@@ -86,3 +94,25 @@ type PullProgressResponse struct {
...
 }
+
+type BeginTransactionRequest struct {
+}
+
+type BeginTransactionResponse struct {
+	TransactionID int `json:"transactionId"`
+}
+
+type CloseTransactionRequest TransactionRequest
```
Comment (aboodman) [inline repm/types.go:105]: Why distinguish between `CloseTransaction` and `Abort`/`Commit`? Seems like `Abort` and `Close` are the same thing.

**Comment by aboodman** [comment]
> Should the cli expose begin transaction?

No. I think the CLI is going to lose its ability to directly edit the cache. It will become read-only.

> Who owns the transactions map? Maybe the connection does?

Connection I think.

> Maybe only api needs this map?

Huh. That's actually pretty interesting. Worth trying.

**Comment by aboodman** [comment]
This is coming along nicely. I didn't go through the logic and naming/etc in detail and focused on structural changes for this round.

File: db/db.go (hunk 1 above)
Comment (arv) [reply db/db.go:386]: Scan is not yet implemented on a `MapEditor`

Comment (arv) [reply db/db.go:386]: I'll look into doing the iterator on MapEditor before continuing.

File: db/db.go (hunk 5 above)
Comment (arv) [reply db/db.go:519]: Agreed. I was keeping it because I wanted to copy over the function name and args code

File: db/db.go (hunk 6 above)
Comment (arv) [reply db/db.go:524]: I was thinking that the failure to FastForward would need to be handled outside `Commit` since we need to replay the whole transaction.

File: repm/types.go (hunk 8 above)
Comment (arv) [reply repm/types.go:105]: Yes. They are the same thing. I was thinking from a classical DB terminology with BEGIN COMMIT/ROLLBACK and it felt strange.

File: db/db.go (hunk 1 above)
Comment (arv) [reply db/db.go:386]: I think the simplest solution is to just build the Map at that point. We do not want to reflect changes to the MapEditor during scan anyway. If this ever becomes a perf issue we can implement a `types.MapIterator` or a smaller subset of that interface for `types.MapEditor`

Comment (aboodman) [reply db/db.go:386]: It's true that `MapIterator` lacks scan, but I don't see the relevance to the question of whether to combine these two structs. Both `ReadTransaction` and `WriteTransaction` require `Scan`.

File: db/db.go (hunk 6 above)
Comment (aboodman) [reply db/db.go:524]: Agree but how is that related? To be concrete, I think we need to do something like this:

```
defer tx.db.lock()()
_, err = tx.noms.FastForward(tx.noms.GetDataset(LOCAL_DATASET), ref)
if err != nil {
	return
}
tx.db.head = commit
return
```

Comment (aboodman) [reply db/db.go:524]: That seems compatible with handling retry errors at the callsite.

File: repm/types.go (hunk 8 above)
Comment (aboodman) [reply repm/types.go:105]: Probably internally databases look more like what's happening here.

File: db/db.go (hunk 6 above)
Comment (arv) [reply db/db.go:524]: That is what I have but I didn't upload the latest snapshot yet.

File: db/db.go (hunk 2 above)
Comment (arv) [reply db/db.go]: Removed alltogether

File: db/db.go (hunk 3 above)
Comment (arv) [reply db/db.go]: Made the connection own them since the mapping is only needed for the rpc

File: db/db.go (hunk 5 above)
Comment (arv) [reply db/db.go]: They are used by `rebase`. Keeping them for now

**Comment by arv** [comment]
PTAL

I will look at the Java/ObjC/Dart parts tomorrow

File: db/db_test.go [hunk 9]
```diff
@@ -20,13 +22,19 @@ func reloadDB(assert *assert.Assertions, dir string) (db *DB) {
 	return db
 }
 
+func assertDataEquals(assert *assert.Assertions, db *DB, expr string) {
+	value := db.head.Value.Data.TargetValue(db.Noms())
+	assert.True(nomdl.MustParse(db.Noms(), expr).Equals(value), "Expected %s equal %s", types.EncodeValue(value), expr)
```
Comment (aboodman) [inline db/db_test.go:27]: You can compare the refs without loading the value:

`assert.True(types.NewRef(nomdl.MustParse(db.Noms(), expr))).Equals(db.head.Value.Data))`

File: db/db_test.go [hunk 10]
```diff
@@ -111,4 +141,197 @@ func TestLoadBadSpec(t *testing.T) {
...
+	act, err = tx2.Get("k")
+	assert.NoError(err)
+	assert.Equal(v2, act, "expected %s got %s", v2, act)
+
+	_, err = tx1.Commit()
+	assert.NoError(err)
+	assertDataEquals(assert, db, `map {"k": "v1"}`)
+
+	_, err = tx2.Commit()
+	assert.Equal("Dataset head is not ancestor of commit", err.Error())
```
Comment (aboodman) [inline db/db_test.go:303]: Can we detect this and have a more clear error, particularly since callee is going to have to retry? (Can be a separate PR).

File: repm/api.go [hunk 11]
```diff
@@ -176,6 +213,73 @@ func (conn *connection) dispatchPullProgress(reqBytes []byte) ([]byte, error) {
...
+	txID := conn.newTransaction()
+
+	res := BeginTransactionResponse{
+		TransactionID: txID,
+	}
+	return mustMarshal(res), nil
+}
+
+func (conn *connection) dispatchCloseTransaction(reqBytes []byte) ([]byte, error) {
+	var req TransactionRequest
```
Comment (aboodman) [inline repm/api.go:241]: For documentation purposes let's just create a `BeginTransactionRequest` that is a renamed `TransactionRequest`

File: cmd/repl/main.go [hunk 12]
```diff
@@ -244,7 +251,23 @@ func put(parent *kingpin.Application, gdb gdb, in io.Reader) {
 		if _, err := v.ReadFrom(in); err != nil {
 			return err
 		}
-		return db.Put(*id, v.Bytes())
+
+		tx := db.NewTransaction()
```
Comment (aboodman) [inline cmd/repl/main.go:255]: There's no need to implement this because the `put` feature of repl must go away (because we'd need a server side endpoint to call to make it meaningful).

File: cmd/repl/main.go [hunk 13]
```diff
@@ -256,14 +279,20 @@ func del(parent *kingpin.Application, gdb gdb, out io.Writer) {
 		if err != nil {
 			return err
 		}
-		ok, err := db.Del(*id)
-		if err != nil {
-			return err
-		}
-		if !ok {
+		tx := db.NewTransaction()
```
Comment (aboodman) [inline cmd/repl/main.go:282]: Same as `put` - this feature of the CLI needs to go away.

File: db/db_test.go [hunk 14]
```diff
@@ -76,24 +93,37 @@ func TestDel(t *testing.T) {
 	db, err := Load(sp)
 	assert.NoError(err)
 
-	err = db.Put("foo", []byte(`"bar"`))
+	wtx := db.NewTransaction()
```
Comment (aboodman) [inline db/db_test.go:96]: I'll leave it to you, but I've really come to appreciate the go-style table driven tests. For things like this that we really care about every branch being tested and working per-design, it's easy to prove to yourself that you've covered them all.

You'd take each individual method of `Transaction` and create a test struct that has a field for each case where the code branches. So for like `Transaction::Put`:

```
func TestTransactionPut(t *testing.T) {
...
tc := []struct {
    txClosed bool
    path string
    json string
    expectedError string
}{
    {true, "foo", "42", ErrorClosed.Error()},
    {false, "", "42", "Empty ID not supported"},
    {false, "foo", "", "Invalid JSON"},
    ...
}
```

Then just create an entry that exercises each branch.

File: repm/api_test.go [hunk 15]
```diff
@@ -48,37 +50,75 @@ func TestBasics(t *testing.T) {
...
+		{"scan", `{"transactionId": 5, "start": {"id": {"value": "foo"}}}`, `[{"id":"foo","value":"bar"},{"id":"foopa","value":"doopa"}]`, ""},
+		{"scan", `{"transactionId": 5, "start": {"id": {"value": "foo", "exclusive": true}}}`, `[{"id":"foopa","value":"doopa"}]`, ""},
+		{"closeTransaction", `{"transactionId":5}`, `{}`, ""},
 
 		// TODO: other scan operators
 	}
 
+	currentTXID := 0
+	re := regexp.MustCompile(`"transactionId":\s*(\d+)`)
+	findTransactionID := func(s string) {
```
Comment (aboodman) [inline repm/api_test.go:94]: Isn't the transaction ID consistent across runs of the test? The state is per-connection so I don't see how it can change and why this string stuff is needed.

File: db/db.go [hunk 16]
```diff
@@ -245,9 +214,142 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+	if ok {
+		err = tx.me.Remove(k)
+		if err == nil {
+			tx.didWrite = true
+		}
+	}
+	return ok, err
+}
+
+func (tx *Transaction) Close() error {
```
[the remark this answers is not in the document]
Comment (aboodman) [reply db/db.go:315]: Nevermind.

**Comment by aboodman** [comment]
> I will look at the Java/ObjC/Dart parts tomorrow

I don't think there will be any Java/Obj-C changes required since we haven't changed the RPC style at all (we don't call back into Dart for example).

After this change, the Java/Obj-C parts will be too conservative -- they use a single queue to dispatch transactions when they could multithread them. But that is a small change and can be separate.

File: db/db.go [hunk 17]
```diff
@@ -245,9 +214,142 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+}
+
+func (tx *Transaction) Scan(opts ScanOptions) ([]ScanItem, error) {
+	if tx.Closed() {
+		return nil, ClosedError
+	}
+	return scan(tx.me.Build().NomsMap(), opts)
+}
+
+func (tx *Transaction) Put(path string, JSON []byte) error {
```
[the remark this answers is not in the document]
Comment (arv) [reply db/db.go:276]: I have both gofmt and golint enabled. It is happy with this but I'm not happy with this

[the remark this answers is not in the document]
Comment (aboodman) [reply db/db.go:276]: Huh, weird. Maybe my understanding of Go style here is wrong. Oh well.

File: db/db.go [hunk 18]
```diff
@@ -245,9 +214,142 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+	if tx.Closed() {
+		return ClosedError
+	}
+
+	canonicalJSON, err := nomsjson.Canonicalize(JSON)
+	if err != nil {
+		return fmt.Errorf("could not Put '%s'='%s': %w", path, JSON, err)
```
[the remark this answers is not in the document]
Comment (arv) [reply db/db.go:283]: It is already using `%w`

https://blog.golang.org/go1.13-errors

File: db/db_test.go (hunk 9 above)
Comment (arv) [reply db/db_test.go:27]: I will do that but if it fails I still want to load it to give a nicer error message.

File: db/db.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (aboodman) [reply db/db.go:283]: 🤯

File: repm/api.go (hunk 11 above)
Comment (arv) [reply repm/api.go:241]: Not sure what you want here?

File: cmd/repl/main.go (hunk 12 above)
Comment (arv) [reply cmd/repl/main.go:255]: Lets remove it when we fix rebase

File: db/db_test.go (hunk 14 above)
Comment (arv) [reply db/db_test.go:96]: This one used to be a lot more subtle when there we read and write transactions.

Comment (arv) [reply db/db_test.go:96]: One thing I really do not like about table driven tests (the way we write them) is that it is hard to see which test failed. I thought that `t.Run` was supposed to help but it doesn't (it does allow running subtests which is nice)

File: repm/api_test.go (hunk 15 above)
Comment (arv) [reply repm/api_test.go:94]: It wasn't before. Let me strip this cruft

File: db/db.go [hunk 19]
```diff
@@ -245,9 +214,142 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
 }
+
+type Transaction struct {
+	db       *DB
+	noms     datas.Database
+	head     Commit
+	me       *kv.MapEditor
+	didWrite bool
+	closed   bool
+	Name     string
```
[the remark this answers is not in the document]
Comment (arv) [reply db/db.go:241]: For now I wanted it mutable but at one point I had an overloaded `NewTransactionWithArgs` function. I will make it that way for now because it is more restrictive.

File: repm/api.go [hunk 20]
```diff
@@ -176,6 +213,73 @@ func (conn *connection) dispatchPullProgress(reqBytes []byte) ([]byte, error) {
...
+	var req closeTransactionRequest
+	err := json.Unmarshal(reqBytes, &req)
+	if err != nil {
+		return nil, err
+	}
+	tx, err := conn.findTransaction(req.TransactionID)
+	if err != nil {
+		return nil, err
+	}
+	err = tx.Close()
```
Comment (phritz) [inline repm/api.go:250]: might be slightly safer to remove it from the map first then close it on the idea that the map holds only open transactions? (same applies to commit below)

File: repm/api.go [hunk 21]
```diff
@@ -176,6 +213,73 @@ func (conn *connection) dispatchPullProgress(reqBytes []byte) ([]byte, error) {
...
+
+	txID := conn.newTransaction()
+
+	res := beginTransactionResponse{
+		TransactionID: txID,
+	}
+	return mustMarshal(res), nil
+}
+
+func (conn *connection) dispatchCloseTransaction(reqBytes []byte) ([]byte, error) {
```
Comment (phritz) [inline repm/api.go:240]: ocd brain nit: begin/close mismatch, begin/end or open/close?

File: db/transaction.go [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+	db     *DB
+	head   Commit
+	me     *kv.MapEditor
+	wrote  bool
+	closed bool
+	name   string
+	args   types.List
+}
+
+// Closed returns true when the database has been closed. A database becomes
```
Comment (phritz) [inline db/transaction.go:31]: possible search and replace misfire? i think you mean transaction has been close and a transaction becomes closed, not a database...?

File: db/transaction.go [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+// closed after Committ or Close is called.
+func (tx *Transaction) Closed() bool {
+	return tx.closed
+}
+
+// Get returns the JSON encoded value at the given id of the database.
+func (tx *Transaction) Get(id string) ([]byte, error) {
```
Comment (phritz) [inline db/transaction.go:38]: fwiw these args are variously called id, path, and key in different places. not sure if it is worth trying to standardize on one.

File: db/transaction.go [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+	}
+	tx.closed = true
+	return nil
+}
+
+// Commit tries to commits the changes made to the database in this transaction.
+// If this returns without an error the commit succeeded and the new ref of the
+// database head is returned.
+func (tx *Transaction) Commit() (ref types.Ref, err error) {
+	if tx.closed {
```
Comment (phritz) [inline db/transaction.go:125]: i feel just a little discomfort that closed is not protected by a mutex. it looks like so long as you switch the order of marking a transaction closed/committed and removing it from the map (see comment elsewhere) we shouldn't ever call commit concurrently, however... might we feel a little safer if we knew it was *impossible* for a transaction to be committed twice (eg by adding a mutex on closed)?

File: db/db.go [hunk 25]
```diff
@@ -245,9 +208,34 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
 	return newData, newDataChecksum, output, isWrite, nil
 }
 
+// NewTransaction returns a new Transaction.
+func (db *DB) NewTransaction() *Transaction {
+	return &Transaction{
+		db:   db,
+		head: db.head,
```
Comment (phritz) [inline db/db.go:215]: i don't think we're using db.head in a safe way:
1. we lock when we write to it but do not lock when we read it
2. we read db.head multiple times in a function and assume that it has not changed eg here in this example (db.head could change between setting head: and setting me:)

seems like we should have (not necessarily as part of this pr, just soon) a getter and a setter for head that both use the lock, functions should read the value exactly once, and all the code should stop fiddling with db.head directly. (this applies in diffserver as well)

cc @aboodman

File: db/db.go [hunk 26]
```diff
@@ -245,9 +208,34 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
...
+	}
+}
+
 func (db *DB) lock() func() {
 	db.mu.Lock()
 	return func() {
 		db.mu.Unlock()
 	}
 }
+
```
Comment (phritz) [inline db/db.go:239]: the code this comment applies to moved

File: db/db.go (hunk 25 above)
Comment (aboodman) [reply db/db.go:215]: The second one is definitely true, good catch, scary. The first one doesn't seem dangerous to me but maybe i'm not seeing it.

File: db/transaction.go (hunk 23 above)
Comment (aboodman) [reply db/transaction.go:38]: *ducks*

I think we should probably standardize on `key` since this is a "key value database". Whoops. This is fine for now.

Comment (aboodman) [reply db/transaction.go:38]: And also because this is a cache and generally the lookup thingy in a cache is called _key_.

Comment (arv) [reply db/transaction.go:38]: I prefer `key` but I don't have a string preference. I agree that we should standardize since it does come up in `ScanItem`

File: db/db.go (hunk 25 above)
Comment (phritz) [reply db/db.go:215]: the first one is definitely true as well: setting a value is not guaranteed to be atomic so setting db.head while someone is reading db.head could has undefined behavior. i think we'd have to use something like https://golang.org/pkg/sync/atomic/#LoadPointer to guarantee a safe read if we are not locking.

File: repm/api.go [hunk 27]
```diff
@@ -12,17 +14,40 @@ import (
...
+	if tx, ok := conn.transactions[txID]; ok {
+		return tx, nil
+	}
+	return nil, fmt.Errorf("Invalid transaction ID: %d", txID)
+}
+
+func (conn *connection) removeTransaction(txID int) {
+	conn.transactionMutex.Lock()
+	defer conn.transactionMutex.Unlock()
+	delete(conn.transactions, txID)
```
[the remark this answers is not in the document]
Comment (arv) [reply repm/api.go:48]: All the callers are already "guarded" by findTransaction.

File: db/db.go (hunk 25 above)
Comment (phritz) [reply db/db.go:215]: https://github.com/rocicorp/replicache/issues/42

File: db/transaction.go (hunk 23 above)
Comment (phritz) [reply db/transaction.go:38]: key is great, let's do it

File: repm/api.go [hunk 28]
```diff
@@ -12,17 +14,40 @@ import (
...
 type pullProgress struct {
 	bytesReceived uint64
 	bytesExpected uint64
 }
 
+func (conn *connection) findTransaction(txID int) (*db.Transaction, error) {
+	if txID == 0 {
+		return nil, fmt.Errorf("Missing transaction ID")
+	}
+	if tx, ok := conn.transactions[txID]; ok {
```
[the remark this answers is not in the document]
Comment (arv) [reply repm/api.go:39]: Agreed.

File: repm/api.go (hunk 21 above)
Comment (arv) [reply repm/api.go:240]: Agreed. Switch to open/close

File: db/transaction.go (hunk 24 above)
Comment (arv) [reply db/transaction.go:125]: I agree that after guarding the maps it cannot happen.

Adding a mutex to ensure that this never cause any problem in the future

File: db/transaction.go [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+	"roci.dev/diff-server/util/time"
+)
+
+var (
+	// ErrClosed is the error returned from operations on a Transaction when
+	// it has already been closed.
+	ErrClosed = errors.New("Transaction is closed")
+)
+
+type Transaction struct {
```
Comment (phritz) [inline db/transaction.go:21]: I think you should make a statement here about the threadsafety of Transaction and I *think* we need to ensure that it is threadsafe. (Even if we have a mechanism in bindings that serializes access, it just seems safer). I think this means:
- guard closed with a mutex
- guard mapeditor operations put/get/has/del with a mutex. (mapeditor is not safe for concurrent access https://github.com/attic-labs/noms/blob/39057233bfdde4b20dbfc8fa8c08a45f0fbe9360/go/types/map_editor.go#L19 -- no locks around access to mapeditor.edits)

cc @aboodman 
-

File: db/transaction.go (hunk 24 above)
Comment (arv) [reply db/transaction.go:125]: Thinking more... I think we might want to guard every method of a tx with a mutex. For example we do not want a `Put` to happen in the middle of a `Commit`.

Comment (phritz) [reply db/transaction.go:125]: yes, that's an even stronger condition than I suggested in https://github.com/rocicorp/replicache-client/pull/60#pullrequestreview-395086277 and i think it makes sense. we will need to be sure we are checking closed inside the lock in case someone else closed or committed while we were waiting.

File: db/transaction.go (hunk 29 above)
Comment (arv) [reply db/transaction.go:22]: I added a mutex around the operations.

I'll update the comment too.

**Comment by arv** [comment]
Everything resolved and bugs filed...

SDK changes here: https://github.com/rocicorp/replicache-sdk-flutter/pull/29

Will land tomorrow

File: db/transaction.go (hunk 29 above)
Comment (aboodman) [reply db/transaction.go:25]: @phritz thanks for thinking about this, you are correct. This is something we need to agree on the design around and document in the class comments of `Transaction` and `db` carefully.

I think it would be reasonable to decide that all access to `repm` is (*must be*) serialized externally, or it would be reasonable to decide that it can be accessed concurrently.

Given the design of Noms and the transaction system we have proposed on top of it in the bug associated with this PR, it **seems** reasonable to me to allow concurrent acccess to `db` and `Transaction`. And if that's what we're doing, @phritz feedbacks are correct.

If either of you have strong reasons why access to db *must be* allowed to be concurrent or *must not be* allowed to be, that would be useful to document.

Comment (phritz) [reply db/transaction.go:25]: i think if we dont have concurrent access to db then all db activity has to stop during sync, which doesn't sound good. no concurrent access to db also has a fishy smell: databases are not usually serial so i never considered that that was the case. my guess is there is there'd be more undesirable side effects if it were. dunno if those are strong reasons.

also i think(?) the existence of the transaction machinery at the level of repm/client implies to me that concurrent access to Transaction is possible. otherwise i think(?) if access is strictly serialized at the bindings layer then that eliminates the need for a lot of the transaction machinery at the repm/client layer.

disclaimer: i'm new
